addons/mrp_subcontracting/models/stock_move.py

```diff
@@ -3,39 +3,141 @@
+                continue
             subcontract_details_per_picking[move.picking_id].append((move, bom))
+            move.write({
+                'is_subcontract': True,
+                'location_id': self.picking_id.partner_id.property_stock_subcontractor.id
+            })
         for picking, subcontract_details in subcontract_details_per_picking.items():
             picking._subcontracted_produce(subcontract_details)
+
+        res = super(StockMove, self)._action_confirm(merge=merge, merge_into=merge_into)
+        if subcontract_details_per_picking:
+            self.env['stock.picking'].concat(*list(subcontract_details_per_picking.keys())).action_assign()
         return res
+
+    def _action_record_components(self):
+        action = self.env.ref('mrp.act_mrp_product_produce').read()[0]
+        action['context'] = dict(
+            default_production_id=self.move_orig_ids.production_id.id,
+            default_subcontract_move_id=self.id
+        )
+        return action
+
+    def _check_overprocessed_subcontract_qty(self):
+        overprocessed_moves = self.env['stock.move']
+        for move in self:
+            if not move.is_subcontract:
+                continue
+            # Extra quantity is allowed when components do not need to be register
+            if not move._has_tracked_subcontract_components():
+                continue
+            rounding = move.product_uom.rounding
+            if float_compare(move.quantity_done, move.move_orig_ids.production_id.qty_produced, precision_rounding=rounding) > 0:
+                overprocessed_moves |= move
+        if overprocessed_moves:
+            raise UserError(_("""
+You have to use 'Records Components' button in order to register quantity for a
+subcontracted product(s) with tracked component(s):
+ %s.
+If you want to process more than initially planned, you
+can use the edit + unlock buttons in order to adapt the initial demand on the
+operations.""") % ('\n'.join(overprocessed_moves.mapped('product_id.display_name'))))
+
+    def _get_subcontract_bom(self):
+        self.ensure_one()
+        bom = self.env['mrp.bom'].sudo()._bom_subcontract_find(
+            product=self.product_id,
+            picking_type=self.picking_type_id,
+            company_id=self.company_id.id,
+            bom_type='subcontract',
+            subcontractor=self.picking_id.partner_id,
+        )
+        return bom
+
+    def _has_tracked_subcontract_components(self):
+        self.ensure_one()
+        return any(m.has_tracking != 'none' for m in self.move_orig_ids.production_id.move_raw_ids)
+
+    def _update_subcontract_order_qty(self):
+        for move in self:
+            production = move.move_orig_ids.production_id
+            if production:
+                self.env['change.production.qty'].with_context(skip_activity=True).create({
+                    'mo_id': production.id,
+                    'product_qty': move.product_uom_qty
+                }).change_prod_qty()
```

Comment:
for all these new methods

Private attributes (_name, _description, _inherit, …)
Default method and _default_get
Field declarations
Compute, inverse and search methods in the same order as field declaration
Selection method (methods used to return computed values for selection fields)
Constrains methods (@api.constrains) and onchange methods (@api.onchange)
CRUD methods (ORM overrides)
Action methods
And finally, other business methods.